src/middleware/withContentClaimsDagula.js

```diff
@@ -46,3 +52,70 @@ export function withContentClaimsDagula (handler) {
     return handler(request, env, { ...ctx, blocks: dagula, dag: dagula, unixfs: dagula })
   }
 }
+
+/**
+ * Returns the cached DAG Protobuf bytes if they exist, otherwise fetches the DAG Protobuf bytes
+ * from the fetcher and caches them in the KV store.
+ *
+ * @param {Environment} env
+ * @param {import('@web3-storage/blob-fetcher').Fetcher} fetcher
+ * @param {import('multiformats').UnknownLink} cid
+ * @returns {Promise<{ cid: import('multiformats').UnknownLink, bytes: Uint8Array } | undefined>}
+ */
+async function getDagPbContent (env, fetcher, cid) {
+  if (env.FF_DAGPB_CONTENT_CACHE_ENABLED === 'true' && cid.code === dagPb.code) {
+    const cachedBytes = await getCachedDagPbBytes(env, cid)
+    if (cachedBytes) {
+      return { cid, bytes: cachedBytes }
+    }
+
+    const res = await fetcher.fetch(cid.multihash)
+    if (res.ok) {
+      const bytes = await res.ok.bytes()
+      const cachedBytes = await cacheDagPbBytes(env, cid, bytes)
+      return { cid, bytes: cachedBytes }
+    }
+  }
+  return undefined
+}
+
+/**
+ * Caches the DAG Protobuf content into the KV store if the content size is less than or equal to the max size.
+ * The content is cached for the duration of the TTL (seconds), if the TTL is not set, the content is cached indefinitely.
+ *
+ * @param {Environment} env
+ * @param {import('multiformats').UnknownLink} cid
+ * @param {Uint8Array} bytes
+ * @returns {Promise<Uint8Array>}
+ */
+async function cacheDagPbBytes (env, cid, bytes) {
+  if (env.FF_DAGPB_CONTENT_CACHE_MAX_SIZE_MB && bytes.length <= env.FF_DAGPB_CONTENT_CACHE_MAX_SIZE_MB * 1024 * 1024) {
+    try {
+      await env.DAGPB_CONTENT_CACHE.put(cid.multihash.digest.toString(), bytes.buffer, {
```

Comment:
Old DAGs will not have "raw" blocks as leaves, they will be `dag-pb` and the UnixFS data (decoded from the `Data` property of the node) will be of type "file". We should not cache these.

<img width="1582" alt="Screenshot 2025-01-24 at 11 28 13" src="https://github.com/user-attachments/assets/aa134f72-9760-420b-8a73-91f13209ce40" />

I think you don't have to decode the unixfs data, you just need to `dagPb.decode(bytes)` and check that `Links` has length 0.

Reply:
Nice. Thank you!